test/metabase/lib/schema/filter_test.cljc

```diff
@@ -0,0 +1,89 @@
+(ns metabase.lib.schema.filter-test
+  (:require
+   [clojure.test :refer [deftest is testing]]
+   [clojure.walk :as walk]
+   [malli.core :as mc]
+   malli.registry
+   ;; expression and filter recursively depend on each other
+   metabase.lib.schema.expression
+   [metabase.lib.schema.filter :as filter]
+   [metabase.util.malli.registry :as mr]))
+
+(defn- ensure-uuids [filter-expr]
+  (walk/postwalk
+   (fn [f]
+     (if (and (vector? f) (not (map-entry? f)))
+       (let [[op & args] f]
+         (cond
+           (and (map? (first args)) (not (:lib/uuid (first args))))
+           (assoc-in f [1 :lib/uuid] (str (random-uuid)))
+
+           (not-any? #(and (map? %) (:lib/uuid %)) args)
+           (into [op {:lib/uuid (str (random-uuid))}] args)
+
+           :else f))
+       f))
+   filter-expr))
+
+(defn- known-filter-ops
+  "Return all registered filter operators."
+  []
+  (into #{}
+        (comp (filter #(and (qualified-keyword? %)
+                            (not= % ::filter/filter)
+                            (= (namespace %) (namespace ::filter/filter))))
+              (map (comp keyword name)))
+        (keys (malli.registry/schemas @#'mr/registry))))
+
+(defn- filter-ops
+  "Return the set of filter operators in `filter-expr`."
+  [filter-expr]
+  (loop [stack [filter-expr] ops #{}]
+    (if (seq stack)
+      (let [top (peek stack)
+            others (pop stack)]
+        (if (and (vector? top)
+                 (keyword? (first top))
+                 (not (#{:field} (first top))))
+          (recur (into others (rest top)) (conj ops (first top)))
+          (recur others ops)))
+      ops)))
+
+(deftest ^:parallel filter-test
+  (testing "valid filters"
+    (let [field [:field 1 {:lib/uuid (str (random-uuid))}]
+          boolean-field [:field 2 {:lib/uuid (str (random-uuid))
+                                   :base-type :type/Boolean}]
```

Comment:
Just a heads up, I moved the `:field` options to the second arg in #29010. If this PR lands first then I can update these tests